File: java/primerosProgramasPOO/Fraccion.java
package primerosProgramasPOO;

/**
 * 
 * 5. Crea una clase Fraccion (en Java yPython) de forma que podamos hacer las siguientes operaciones:
 * - Contruir un objeto Fraccion pasándole al constructor el numerador y el denominador.
 * - Obtener la fracción.
 * - Obtener y modificar numerador y denominador. No se puede dividir por cero.
 * - Obtener resultado de la fracción (número real).
 * - Multiplicar la fracción por un número.
 * - Multiplicar, sumar y restar fracciones.
 * - Simplificar la fracción.
 * 
 * @author Javier Lopera Jimenez
 *
 */

import javax.swing.JOptionPane;

public class Fraccion {

	double numerador;
	double denominador;
	
	//Constructor de Fraccion (el denominador no puede valer 0)
	
	Fraccion(double numerador,double denominador){
		
		if(denominador==0) {
			
			denominador=Double.parseDouble(JOptionPane.showInputDialog("El denominador no puede ser 0, introduce un nuevo denominador:"));
			
		}
		
		
		this.numerador=numerador;
		this.denominador=denominador;
				
	}
	
	//Getter del numerador
	
	public double getNumerador() {
		
		return this.numerador;
		
	}
	
	//Getter del denominador
	
	public double getDenominador() {
		
		return this.denominador;
	
	}
	
	//Setter del numerador
	
	private double setNumerador(double numerador) {
		
		this.numerador=numerador;
		
		return numerador;
		
	}
	
	//Setter del denominador, ademas el denominador no puede valer 0
	
	private double setDenominador(double denominador) {
		
		if(denominador==0) {

			denominador=Double.parseDouble(JOptionPane.showInputDialog("El denominador no puede ser 0, introduce un nuevo denominador:"));
		
		}
		
		
		this.denominador=denominador;
		
		return denominador;
		
	}
	
	

	public String modificarFraccion(int nuevoNumerador, int nuevoDenominador) {
		
		this.setNumerador(nuevoNumerador);
		
		this.setDenominador(nuevoDenominador);
		
		return toString();
		
		
	}
	
	public double resultadoFraccion() {
		
		double total=getNumerador()/getDenominador();
		
		return total;
		
	}
	
	public String multiplicar(double numero) {
		
		this.numerador=numerador*numero;
		
		return toString();
		
	}
	
	public Fraccion multiplicarPorFraccion(Fraccion fraccion2) {
		
		return new Fraccion(numerador*fraccion2.getNumerador(),denominador*fraccion2.getDenominador());
		
	}
		
	public Fraccion dividirPorFraccion(Fraccion fraccion2) {
		
		return new Fraccion(numerador*fraccion2.getDenominador(),denominador*fraccion2.getNumerador());
		
	}
	
/*	
	public Fraccion sumar(Fraccion denominador2) {
		
		double denominadorComun = mcm(getDenominador(),denominador2.getDenominador());
		
		resultadoNumerador= ;
		
		return ;
	}
*/
	
	public Fraccion simplifica() {
  
    double n = this.numerador;
    double d = this.denominador;
      
    for (int i = 2; i < Math.min(this.numerador, this.denominador); i++) {
      while (((n % i) == 0) && ((d % i) == 0)) {
        n /= i;
        d /= i;
      }
    }

    return new Fraccion(n, d);
  }

	public double mcm(double a, double b) {
		
		double denominador;
		
		if(a>b)
			denominador=a;
		else
			denominador=b;
		
		while (denominador%a!=0 || denominador%b!=0)
			denominador++;
		
		return denominador;
		
	}
	
	public Fraccion sumar(Fraccion f) {
		
		double mcm = mcm(this.denominador, f.denominador);
		
		double nuevoNumerador1 =  (mcm/this.denominador)*this.numerador;
		double nuevoNumerador2 =  (mcm/f.denominador)*f.numerador;
		double numeradorFinal = nuevoNumerador1+nuevoNumerador2;

		return new Fraccion(numeradorFinal,mcm);
	}
	
	public Fraccion restar(Fraccion f) {

		double mcm = mcm(this.denominador, f.denominador);
		
		double nuevoNumerador1 =  (mcm/this.denominador)*this.numerador;
		double nuevoNumerador2 =  (mcm/f.denominador)*f.numerador;
		double numeradorFinal = nuevoNumerador1-nuevoNumerador2;
		
		return new Fraccion(numeradorFinal,mcm);
	}
	
	@Override
	public String toString() {
		return numerador + "/" + denominador;
	}
	
	
	
	
}

File: python/primerosProgramasPOO/Fraccion.py
'''
Created on 22 ene. 2019

@author: d18lojij
'''

class Fraccion:
	def __init__(self,n,d):
		self.n=n
		self.d=d
		
	def setNumerador(self,n):
		self.n=n
		return n
	
	def setDenominador(self,d):
		self.d=d
		return d
	
	def modificarFraccion(self,n,d):
		self.n=n
		self.d=d
		return str(self.n)+"/"+str(self.d)
	
	def resultadoFraccion(self):
		total=self.n/self.d
		return total
	
	def multiplicar(self,numero):
		self.n=self.n*numero
		return str(self.n)+"/"+str(self.d)
	
	def multiplicarPorFraccion(self,fraccion):
		
		return Fraccion(self.n*fraccion.n,self.d*fraccion.d)
	
	
		
	
	def __str__(self):
		return str(self.n)+"/"+str(self.d)
	

if __name__ == '__main__':
	
	f1=Fraccion(2.5,3)
	f2=Fraccion(3,4)
	
	print("La fraccion inicial es : " , f1)
	print("Resultado de la fraccion : " , f1.resultadoFraccion())
	f1.modificarFraccion(2,5)
	print("Modificamos la fraccion: " , f1)
	print("Obtenemos el numerador: " , f1.n)
	print("Obtenemos el denominador: " , f1.d)
	print("Resultado de la nueva fraccion : " , f1.resultadoFraccion())
	print("Multiplicamos la fraccion por 8: " , f1.multiplicar(8))
	
	f3=f1.multiplicarPorFraccion(f2)
	
	print("La fraccion producto de las dos primeras es: " , f3)
	
	
	
	
	
	
	
	
	